File: packages/typescript/src/serve.ts
import { readFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import type { IncomingMessage, ServerResponse } from "node:http";
import type { McpSpec } from "./types.js";
import { serializeSpec } from "./generate.js";

const currentDir = dirname(fileURLToPath(import.meta.url));

function loadDocsTemplate(): string {
  try {
    return readFileSync(join(currentDir, "ui/docs.html"), "utf-8");
  } catch {
    return "<!DOCTYPE html><html><head><title>__TITLE__</title></head><body><pre>__SPEC_DATA__</pre></body></html>";
  }
}

const docsTemplate = loadDocsTemplate();

function buildDocsHtml(spec: McpSpec): string {
  const title = spec.info?.title ?? spec.info?.name ?? "MCP Server";
  return docsTemplate
    .replace("__SPEC_DATA__", JSON.stringify(spec))
    .replace("__TITLE__", title);
}

export interface HandleRequestOptions {
  getSpec: () => McpSpec | null;
  basePath?: string;
}

/**
 * Route handler for mcpspec's built-in routes (/docs, /mcpspec.yaml).
 * Designed to be called from an http.Server request listener.
 * Returns `true` if the request was handled, `false` if it was not
 * a recognized route (caller can then respond with 404 or delegate).
 */
export function handleRequest(
  req: IncomingMessage,
  res: ServerResponse,
  options: HandleRequestOptions,
): boolean {
  const prefix = (options.basePath ?? "").replace(/\/$/, "");
  const url = req.url ?? "/";

  if (req.method === "GET" && url === `${prefix}/docs`) {
    return serveDocs(res, options.getSpec);
  }

  if (req.method === "GET" && url === `${prefix}/mcpspec.yaml`) {
    return serveYaml(res, options.getSpec);
  }

  res.writeHead(404, { "content-type": "text/plain; charset=utf-8" });
  res.end("Not Found");
  return false;
}

function serveDocs(
  res: ServerResponse,
  getSpec: () => McpSpec | null,
): boolean {
  const spec = getSpec();
  if (!spec) {
    res.writeHead(503, { "content-type": "text/plain; charset=utf-8" });
    res.end("Spec not yet available. Server is still initializing.");
    return true;
  }
  const html = buildDocsHtml(spec);
  res.writeHead(200, { "content-type": "text/html; charset=utf-8" });
  res.end(html);
  return true;
}

function serveYaml(
  res: ServerResponse,
  getSpec: () => McpSpec | null,
): boolean {
  const spec = getSpec();
  if (!spec) {
    res.writeHead(503, { "content-type": "text/plain; charset=utf-8" });
    res.end("Spec not yet available. Server is still initializing.");
    return true;
  }
  res.writeHead(200, { "content-type": "text/yaml; charset=utf-8" });
  res.end(serializeSpec(spec));
  return true;
}

File: packages/python/src/mcpspec_dev/serve.py
"""Starlette route handlers for /docs and /mcpspec.yaml endpoints."""

from __future__ import annotations

import html
import json
from collections.abc import Awaitable, Callable
from pathlib import Path
from typing import TYPE_CHECKING

from starlette.responses import PlainTextResponse, Response

from mcpspec_dev.generate import serialize_spec

if TYPE_CHECKING:
    from starlette.requests import Request

    from mcpspec_dev.types import McpSpecDocument

_DOCS_TEMPLATE_PATH = Path(__file__).parent / "ui" / "docs.html"

_FALLBACK_TEMPLATE = (
    "<!DOCTYPE html><html><head><title>__TITLE__</title></head>"
    "<body><pre>__SPEC_DATA__</pre></body></html>"
)


def _load_docs_template() -> str:
    try:
        return _DOCS_TEMPLATE_PATH.read_text(encoding="utf-8")
    except OSError:
        return _FALLBACK_TEMPLATE


_docs_template = _load_docs_template()


def _build_docs_html(spec: McpSpecDocument) -> str:
    """Render the docs HTML page with spec data injected."""
    title = spec.info.title or spec.info.name or "MCP Server"
    json_payload = json.dumps(
        spec.model_dump(by_alias=True, exclude_none=True),
    ).replace("</", "<\\/")
    return _docs_template.replace("__SPEC_DATA__", json_payload).replace(
        "__TITLE__", html.escape(title)
    )


GetSpec = Callable[[], "McpSpecDocument | None"]

SERVICE_UNAVAILABLE_MSG = "Spec not yet available. Server is still initializing."


def create_docs_route(
    get_spec: GetSpec,
) -> Callable[..., Awaitable[Response]]:
    """Factory that returns a Starlette endpoint serving the HTML docs page."""

    async def docs_endpoint(request: Request) -> Response:
        spec = get_spec()
        if spec is None:
            return PlainTextResponse(SERVICE_UNAVAILABLE_MSG, status_code=503)
        return Response(
            content=_build_docs_html(spec),
            media_type="text/html; charset=utf-8",
        )

    return docs_endpoint


def create_yaml_route(
    get_spec: GetSpec,
) -> Callable[..., Awaitable[Response]]:
    """Factory that returns a Starlette endpoint serving the YAML spec."""

    async def yaml_endpoint(request: Request) -> Response:
        spec = get_spec()
        if spec is None:
            return PlainTextResponse(SERVICE_UNAVAILABLE_MSG, status_code=503)
        return Response(
            content=serialize_spec(spec),
            media_type="text/yaml; charset=utf-8",
        )

    return yaml_endpoint
